File: dsa/data_structures/CircularQueue/circularqueue.cpp
#include <iostream>
#include <climits>
using namespace std;

class CircularQueue
{
private:
    int front;
    int rear;
    int *arr;
    int size;

public:
    CircularQueue(int size)
    {
        cout << "Constructor called!" << endl;
        front = -1;
        rear = -1;
        this->size = size;
        arr = new int[size]; // array size determined at run-time
    }

    ~CircularQueue()
    {
        cout << "Destructor called!" << endl;
        delete[] arr;
    }

    bool isFull()
    {
        return (front == 0 && rear == size - 1) || (front == rear + 1); // returns true if any one or both is true
    }

    bool isEmpty()
    {
        return front == -1; // returns true if this is true
    }

    void enqueue(int data)
    {
        if (isFull())
        {
            cout << "CircularQueue is full!" << endl;
        }
        else
        {
            if (front == -1)
            {
                front = 0;
            }
            rear = (rear + 1) % size;
            arr[rear] = data;
            cout << "Enqueued: " << data << endl;
        }
    }

    int dequeue()
    {
        if (isEmpty())
        {
            return INT_MIN;
        }
        else
        {
            int data = arr[front];
            if (front == rear)
            {
                front = rear = -1;
            }
            else
            {
                front = (front + 1) % size;
            }
            return data;
        }
    }

    void display()
    {
        if (isEmpty())
        {
            cout << "CircularQueue is empty!" << endl;
        }
        else
        {
            int i = front;
            do
            {
                cout << arr[i] << " ";
                i = (i + 1) % size;
            } while (i != (rear + 1) % size);
            cout << endl;
        }
    }
};

void circularQueueOperations(CircularQueue &);

int main()
{
    CircularQueue q1(5);
    circularQueueOperations(q1);
    return 0;
}

void circularQueueOperations(CircularQueue &myObj)
{
    int choice;
    int popped;

    cout << "1- enqueue, 2- dequeue, 3- Display, 4- EXIT" << endl;
    do
    {
        cout << "Enter choice: ";
        cin >> choice;

        switch (choice)
        {
        case 1:
            int ele;
            cout << "Enter data to be inserted: ";
            cin >> ele;
            myObj.enqueue(ele);
            break;
        case 2:
            popped = myObj.dequeue();
            if (popped != INT_MIN)
            {
                cout << "Dequeued: " << popped << endl;
            }
            else
            {
                cout << "CircularQueue is empty!" << endl;
            }
            break;
        case 3:
            myObj.display();
            break;
        case 4:
            return;
        default:
            cout << "Wrong Choice!" << endl;
            break;
        }
    } while (1);
}

File: dsa/data-structures/CircularQueue.java
import java.util.Scanner;

class CircularQueueImplementation {
  private int front;
  private int rear;
  private int[] queue;
  private int size;

  CircularQueueImplementation(int size) {
    front = -1;
    rear = -1;
    this.size = size;
    queue = new int[size];
  }


  boolean isEmpty() {
    return front == -1;
  }

  boolean isFull() {
    return front == (rear + 1) % size;
  }

  void enqueue(int val) {
    if (isFull()) {
      System.out.println("Queue is already full");
      return;
    }

    if (front == -1) front = 0;
    rear = (rear + 1) % size;
    queue[rear] = val;
  }

  void dequeue() {
    if (isEmpty()) {
      System.out.println("Queue is already empty");
      return;
    }

    if (front == rear) front = rear = -1;
    front = (front + 1) % size;
  }

  void display() {
    if (isEmpty()) {
      System.out.println("Queue is already empty");
      return;
    }

    for (int i = front; i != rear; i = (i + 1) % size) {
      System.out.print(queue[i] + " ");
    }
    System.out.println();
  }
}

class CircularQueue {
  public static void main(String[] args) {
    CircularQueueImplementation cq = new CircularQueueImplementation(5);

    Scanner sc = new Scanner(System.in);
    System.out.println("1- Enqueue, 2- Dequeue, 3- Print, 4- Exit");

    while(true) {
      System.out.println();
      System.out.print("Enter choice: ");
      int choice = sc.nextInt();

      switch(choice) {
        case 1:
          System.out.print("Enter val: ");
          int val = sc.nextInt();
          cq.enqueue(val);
          break;

        case 2:
          cq.dequeue();
          break;

        case 3:
          cq.display();
          break;

        case 4:
          System.exit(0);
          break;
        
        default:
          System.out.println("Wrong Choice!");
          break;
      }
    }
  }
}